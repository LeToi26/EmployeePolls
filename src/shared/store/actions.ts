import { InitialState, Questions, User, Users } from "shared/model/poll";
import productApi from "../api/productApi";
import { Dispatch } from "@reduxjs/toolkit";

export const RECEIVE_USERS = "RECEIVE_USERS";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

export const SET_CURRENT_USER = "SET_CURRENT_USER";
export const LOGOUT_CURRENT_USER = "LOGOUT_CURRENT_USER";

export const CREATE_QUESTION = "CREATE_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";

export const ADD_ANSWER_USER = "ADD_ANSWER_USER";
export const CREATE_QUESTION_USER = "CREATE_QUESTION_USER";

export function initialData() {
  return (dispatch: Dispatch) => {
    return productApi
      .getInitialData()
      .then(({ users, questions }: { users: any; questions: any }) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
      });
  };
}

export function receiveUsers(users: Users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function receiveQuestions(questions: Questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function handleLogout() {
  return (dispatch: Dispatch) => {
    return dispatch(logoutCurrentUser());
  };
}

export function logoutCurrentUser() {
  return {
    type: LOGOUT_CURRENT_USER,
  };
}

export function handleLogin(username: string, password: string) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();
    const user: any = Object.values(State.users).find(
      (user: any) => user.id === username && user.password === password
    );
    if (!!user) {
      return dispatch(setCurrentUser(user));
    }
  };
}

export function setCurrentUser(authedUser: User | {}) {
  return {
    type: SET_CURRENT_USER,
    authedUser,
  };
}

export function handleAddAnswer(questionId: string, answer: string) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();
    return productApi
      .saveQuestionAnswer(State.authedUser.id, questionId, answer)
      .then(() => {
        dispatch(addAnswerQuestion(State.authedUser.id, questionId, answer));
        dispatch(addAnswerUser(State.authedUser.id, questionId, answer));
      });
  };
}

function addAnswerQuestion(author: string, qid: string, answer: string) {
  return {
    type: ADD_ANSWER_QUESTION,
    author,
    qid,
    answer,
  };
}

export function addAnswerUser(authedUser: string, qid: string, answer: string) {
  return {
    type: ADD_ANSWER_USER,
    authedUser,
    qid,
    answer,
  };
}

export function handleCreateQuestion(
  firstOption: string,
  secondOption: string
) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();

    return productApi
      .saveQuestion(firstOption, secondOption, State.authedUser)
      .then((question) => {
        dispatch(createQuestion(question));
        dispatch(createQuestionUser(question));
      });
  };
}

export function createQuestionUser(question: any) {
  const { author, id } = question;
  return {
    type: CREATE_QUESTION_USER,
    author,
    qid: id,
  };
}

function createQuestion(question: any) {
  return {
    type: CREATE_QUESTION,
    question,
  };
}
