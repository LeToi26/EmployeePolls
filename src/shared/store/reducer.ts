import { InitialState } from "shared/model/poll";
import {
  SET_CURRENT_USER,
  LOGOUT_CURRENT_USER,
  RECEIVE_USERS,
  RECEIVE_QUESTIONS,
  CREATE_QUESTION,
  CREATE_QUESTION_USER,
  ADD_ANSWER_USER,
  ADD_ANSWER_QUESTION,
} from "./actions";

const initialState: InitialState = {
  authedUser: null,
  users: {},
  questions: {},
};

export default function State(state = initialState, action: any) {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        ...state,
        authedUser: action.authedUser,
      };
    case LOGOUT_CURRENT_USER:
      return {
        ...state,
        authedUser: null,
      };
    case RECEIVE_USERS:
      return {
        ...state,
        users: action.users,
      };
    case RECEIVE_QUESTIONS:
      return {
        ...state,
        questions: action.questions,
      };
    case ADD_ANSWER_USER:
      return {
        ...state,
        users: {
          ...state.users,
          [action.authedUser]: {
            ...state.users[action.authedUser as keyof typeof state],
            answers: {
              ...state.users[action.authedUser as keyof typeof state].answers,
              [action.qid]: action.answer,
            },
          },
        },
      };
    case ADD_ANSWER_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.qid]: {
            ...state.questions[action.qid as keyof typeof state],
            [action.answer]: {
              ...state.questions[action.qid as keyof typeof state][
                action.answer
              ],
              votes: state.questions[action.qid as keyof typeof state][
                action.answer
              ].votes.concat(action.author),
            },
          },
        },
      };
    case CREATE_QUESTION_USER:
      return {
        ...state,
        users: {
          ...state.users,
          [action.author]: {
            ...state.users[action.author as keyof typeof state],
            questions: state.users[
              action.author as keyof typeof state
            ].questions.concat(action.qid),
          },
        },
      };
    case CREATE_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.question.id]: action.question,
        },
      };
    default:
      return state;
  }
}
