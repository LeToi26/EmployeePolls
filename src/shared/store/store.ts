import { configureStore } from "@reduxjs/toolkit";
import State from "./reducer";

export const store = configureStore({
  reducer: {
    State,
  },
});
