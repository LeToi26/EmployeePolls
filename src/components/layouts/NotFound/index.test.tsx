/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/render-result-naming-convention */

import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "shared/store/store";
import { BrowserRouter } from "react-router-dom";
import NotFound from "./index";

describe("NotFound", () => {
  it("should render the component when access wrong path", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <NotFound />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });
});
