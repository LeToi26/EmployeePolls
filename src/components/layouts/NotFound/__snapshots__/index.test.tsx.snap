// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFound should render the component when access wrong path 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-gradient-to-r from-purple-300 to-blue-200"
      >
        <div
          class="w-9/12 m-auto py-16 min-h-screen flex items-center justify-center"
        >
          <div
            class="bg-white shadow overflow-hidden sm:rounded-lg pb-8"
          >
            <div
              class="border-t border-gray-200 text-center pt-8"
            >
              <h1
                class="text-9xl font-bold text-purple-400"
              >
                404
              </h1>
              <h1
                class="text-6xl font-medium py-8"
              >
                oops! Page not found
              </h1>
              <p
                class="text-2xl pb-8 px-12 font-medium"
              >
                Oops! The page you are looking for does not exist. It might have been moved or deleted.
              </p>
              <a
                href="/"
              >
                <button
                  class="bg-gradient-to-r from-purple-400 to-blue-500 hover:from-pink-500 hover:to-orange-500 text-white font-semibold px-6 py-3 rounded-md mr-6"
                >
                  HOME
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-gradient-to-r from-purple-300 to-blue-200"
    >
      <div
        class="w-9/12 m-auto py-16 min-h-screen flex items-center justify-center"
      >
        <div
          class="bg-white shadow overflow-hidden sm:rounded-lg pb-8"
        >
          <div
            class="border-t border-gray-200 text-center pt-8"
          >
            <h1
              class="text-9xl font-bold text-purple-400"
            >
              404
            </h1>
            <h1
              class="text-6xl font-medium py-8"
            >
              oops! Page not found
            </h1>
            <p
              class="text-2xl pb-8 px-12 font-medium"
            >
              Oops! The page you are looking for does not exist. It might have been moved or deleted.
            </p>
            <a
              href="/"
            >
              <button
                class="bg-gradient-to-r from-purple-400 to-blue-500 hover:from-pink-500 hover:to-orange-500 text-white font-semibold px-6 py-3 rounded-md mr-6"
              >
                HOME
              </button>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
